ENH: Support sorting frames by a combo of columns and index levels (GH 14353)

jonmmease opened the pull request:
This PR implements the changes proposed in #14353. @jorisvandenbossche 

- [x] closes #14353
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry
- [x] Rework to use utility methods introduced in #17484 after it is merged

Commit:
Support sorting frames by a combo of columns and index levels GH 14353

gfyoung commented on doc/source/basics.rst:
```diff
@@ -1726,8 +1726,9 @@ Sorting
 
-There are two obvious kinds of sorting that you may be interested in: sorting
-by label and sorting by actual values.
+There are three obvious kinds of sorting that you may be interested in: sorting
```
I know this word was in the original doc, but I'm a little uneasy about judgmental words like "obvious."  I think we can just remove it.

jmmease replied:
Agreed, I've reworded it.

gfyoung commented on doc/source/basics.rst:
```diff
@@ -1776,6 +1783,34 @@ argument:
 
+By Indexes and Values
+~~~~~~~~~~~~~~~~~~~~~
+.. versionadded:: 0.21
```
Not sure you need this version-added tag here since you add it below.

jmmease replied:
I left this versionadded tag and removed the one below in the note. This will help people quickly realize that the feature is new and it should be clear that the note below wouldn't apply without the feature.

gfyoung commented on pandas/core/generic.py:
```diff
@@ -67,7 +67,8 @@
         by : str or list of str
-            Name or list of names which refer to the axis items.""")
+            Name or list of names which refer to the axis items or index
+            levels.""")
```
*Not blocking*: My OCD really wants you to not "orphan" the "levels" word at the end of this sentence :smile: .  However, that's largely a me problem, but if you do happen to come up with a wording that doesn't do this orphaning, by all means go for it.

jmmease replied:
Heh, now it bothers me too! I reworked it to fit on one line. "Name or list of names matching axis items or index levels."

Commit:
Documentation cleanup for review

Commit:
Added whatsnew entry

jreback commented:
@jmmease this looks like an older version of #17484

jmmease commented:
@jreback This PR was for the sort_values col&index change. It had some shared utility code with #17484  (the merge col&index change) so i was holding off on updating it until that one was merged.  

Do you want to keep this PR closed until #17484 is merged and I make the updates?

jreback commented:
np can rebase and evaluate after #17484

Commit:
Merge remote-tracking branch 'upstream/master' into enh_14353

# Conflicts:
#	doc/source/basics.rst
#	doc/source/whatsnew/v0.21.0.txt
#	pandas/core/frame.py

Commit:
Moved whatsnew to 0.22.0

Commit:
Convert to new _get_label_or_level_values utility method
Update some documentation vocabulary to label/level distinction

Commit:
Fix up sorting test suite

Commit:
Refactor new test cases.
  - Move to test_sort_values_level_as_str.py
  - Use fixtures and parametrize
  - Added axis=1 test cases

Commit:
Documentation cleanup.

jmmease commented:
Ok, I just pushed a bunch of updates:
- Converted to use the new `_get_label_or_level_values` utility method introduced in #17484
- Moved tests to a new file, updated to use fixtures/parametrize, and added `axis=1` test cases
- I reviewed the `sort_values` documentation in the docstrings and in `basics.rst` and made small language updates throughout.  I hope these changes make things a bit clearer and more consistent.
- I added a full whatsnew section for 0.22.0

@jreback @jorisvandenbossche @gfyoung @TomAugspurger

jmmease commented:
Just following up on this. 

@gfyoung if you have time to look this over at some point, I'd appreciate your feedback on these documentation/docstring updates.

@jreback does this approach look alright to you?

Thanks all

jreback commented:
@jmmease lgtm. @TomAugspurger @jorisvandenbossche if you'd have a look.

jorisvandenbossche commented on pandas/core/frame.py:
```diff
@@ -113,7 +113,15 @@
     optional_by="""
         by : str or list of str
-            Name or list of names which refer to the axis items.""",
+            Name or list of names matching axis levels or off-axis labels.
```
I find "off-axis labels" very complex. I mean I know what you mean and theoretically it is correct, but I think novice users will not understand this (the previous "axis items" was also not that good). 

I am still thinking of a better wording (the annoying thing is that you cannot just say "column labels or index level names" which is probably the common case how this is used)

TomAugspurger replied:
How about "Name or list of names to sort by"? And then things are clarified below.

jmmease replied:
Sounds good. Going with "Name or list of names to sort by"

jorisvandenbossche commented on pandas/tests/frame/test_sort_values_level_as_str.py:
```diff
@@ -0,0 +1,122 @@
+
+    # Sorting by 'idx' should sort by the idx column and raise a
+    # FutureWarning
+    with tm.assert_produces_warning(FutureWarning, check_stacklevel=False):
```
Can you see if you can get this working without `check_stacklevel=False` ?

jmmease replied:
Sure.  Could you offer any insight into what `check_stacklevel` is actually checking? I've had trouble understanding that @jorisvandenbossche

jorisvandenbossche replied:
See https://github.com/pandas-dev/pandas/issues/9584 and https://github.com/pandas-dev/pandas/pull/10676. 
The `stacklevel` is specified in the FutureWarning in `_check_label_or_level_ambiguity`, and is currenlty set at `2`, which is not correct. This should be higher, depending on how many call steps there are between `sort_values` and calling that function. 
The main problem here will probably be that  `_check_label_or_level_ambiguity` is used in multiple places, so the needed stacklevel might differ for the different usages.

jmmease replied:
Thanks, that's helpful

TomAugspurger replied:
The goal is to report the users's line of code that triggered the warning. In our code, when you emit the `FutureWarning` there's an optional `stacklevel` parameter that's controls which line of code the error message is reported with. It's supposed to be the number of function calls from user code to our method emitting the warning.

It's not always possible to get that correct, if you have multiple pandas methods calling the method emitting the warning.

jmmease replied:
Ok, I was able to remove `check_stacklevel=False` for these tests, and for the related tests for `groupby` and `merge`.  I added a `stacklevel` parameter to `_check_label_or_level_ambiguity` and `_get_label_or_level_values` that behaves like the `stacklevel` parameter of `warnings.warn` itself (`stacklevel=1` blames the caller, `stacklevel=2` blames the callers parent, etc.).

Does this look like a good way to handle it? @TomAugspurger @jorisvandenbossche

jorisvandenbossche commented on pandas/tests/frame/test_sorting.py:
```diff
@@ -455,38 +455,38 @@ def test_sort_index_duplicates(self):
         df = DataFrame(np.random.randn(4, 2),
                        columns=MultiIndex.from_tuples([('a', 0), ('a', 1)]))
-        with tm.assert_raises_regex(ValueError, 'levels'):
+        with tm.assert_raises_regex(ValueError, 'not unique'):
```
I think for this case the old error message was more informative, so it would be nice if we could keep a separate message for the case of incomplete key for MI compared to just non-unique key

jmmease replied:
Ok, sure. The new error message will read as follows (for the `MultiIndex` case):
```
ValueError: The column label 'a' is not unique.
For a multi-index, the label must be a tuple with elements corresponding to each level.
```
Does that sound clear? @jorisvandenbossche

jorisvandenbossche commented on pandas/core/frame.py:
```diff
@@ -3636,17 +3640,8 @@ def sort_values(self, by, axis=0, ascending=True, inplace=False,
-                if isinstance(self.columns, MultiIndex):
-                    raise ValueError('Cannot sort by column %s in a '
-                                     'multi-index you need to explicitly '
-                                     'provide all the levels' % str(by))
```
See comment below as well, so it would be nice to somehow keep this message

jorisvandenbossche reviewed (commented):
@jmmease Added some comments!

jmmease commented on doc/source/basics.rst:
```diff
@@ -1738,19 +1738,26 @@ description.
 
-There are two obvious kinds of sorting that you may be interested in: sorting
-by label and sorting by actual values.
+Pandas supports three kinds of sorting: sorting by index levels,
```
Maybe this isn't correct. But I was trying to be consistent in using `level` to refer to the names of the indexes or columns (`df.index.names`, `df.columns.names`) and `label` to refer to the values referenced by the levels (`df.index.get_level_values(lv1)`, `df.columns.get_level_values(lv1)`). Is that in line with accepted vocabulary? @jorisvandenbossche

jorisvandenbossche replied:
Personally I think "index level" is the equivalent of "column", and from there it would follow that the equivalent of "column values" is "index labels of one level / index level labels / index level values" (but none of those sound that good) and the equivalent of "column names/labels" would then be "index level name". 
But it's not that we do that very consistently currently, or that this is necessarily agreed upon vocabulary. But I think it is in line more or less with what you say?

The main reason I commented it here, is because the next line uses "column values", and as you say "label to refer to the values" then to be consistent it would be "index labels". 
And a second reason I commented is that I think "index level" is more complicated, as then a user needs to be familiar with the concept of MultiIndexes, while for this section about sorting, that is not needed.

jmmease replied:
Thanks for clarifying.  After looking at it again, I do agree that "index labels" is the better parallel to "columns values".

TomAugspurger commented on doc/source/basics.rst:
```diff
@@ -1788,6 +1797,36 @@ argument:
+.. versionadded:: 0.22.0
+
+Strings passed as the ``by`` parameter to :meth:`DataFrame.sort_values` may
+refer to either columns or index levels.
```
Same issue here with "index levels". Does changing it to "index names" make sense here, since that's what you're referring to?

Side-question, does this work on a regular (non-multiindex) index?

TomAugspurger replied:
I see you use "index level names" in the whatsnew. That seems like a good option, if you could use it here too.

jmmease replied:
Changed to "index level names". 

And yes, this does work for both the `Index` and `MultIndex` cases.
@TomAugspurger

TomAugspurger commented on pandas/core/frame.py:
```diff
@@ -113,7 +113,15 @@
+            - if `axis` is 1 or `'columns'` then `by` may contain column
+              levels and/or index labels
+
+            Support for specify index/column levels was added in
```
This could be a `..versionmodified:`

    .. versionmodified:: 0.22.0
       Allow specifying index or column level names.

jmmease replied:
done

Commit:
Documentation updates per review

Commit:
Improve warning stacklevel handling for all uses of lable/level utils

Commit:
Reintroduce specialized error message for the multi-index case

jmmease commented on doc/source/basics.rst:
```diff
@@ -1760,20 +1767,22 @@ labels (indexes) are the ``Series.sort_index()`` and the ``DataFrame.sort_index(
-which will use an arbitrary vector or a column name of the DataFrame to
-determine the sort order:
+The :meth:`Series.sort_values` and :meth:`DataFrame.sort_values` methods are
+used to sort a pandas object by its values.  The optional ``by`` parameter to
```
I split this into two sentences so that a `Series` is sorted by "values" and a `DataFrame` is sorted by "column or row values".

jmmease commented on doc/source/basics.rst:
```diff
@@ -1788,6 +1797,36 @@ argument:
+                           index=idx)
+   df_multi
+
+   # Sort by 'second' (index) and 'A' (column)
```
Sounds good

Commit:
Merge remote-tracking branch 'upstream/master' into enh_14353

jmmease commented:
Thanks for the thorough comments @jorisvandenbossche and @TomAugspurger 

I just pushed an update that I believe address all of them.  Let me know if I missed anything or if you see any other updates you'd like before this is merged.

jmmease commented:
@TomAugspurger @jorisvandenbossche Did these latest updates address your comments? Are there any other changes you'd like to see before this is merged? Thanks!

cc @jreback

jmmease commented:
Friendly ping
@TomAugspurger @jorisvandenbossche @jreback

Commit:
Merge branch 'master' of https://github.com/pandas-dev/pandas into enh_14353

Move whatsnew to 0.23.0

# Conflicts:
#	doc/source/whatsnew/v0.22.0.txt

jmmease commented:
@jreback @jorisvandenbossche @TomAugspurger  I noticed the milestone tag for this was moved to 0.23.0.  I just moved the whatsnew to this version. Are we in a freeze for merging things into 0.23.0 until 0.22.0 is released?

Commit:
Update release version references to 0.23.0

Commit:
Merge branch 'master' of https://github.com/pandas-dev/pandas into enh_14353

# Conflicts:
#	doc/source/basics.rst

jreback commented:
@jmmease 0.22 is a special release coming out today. so all previous 0.22 is now 0.23.

jreback commented:
@jmmease just a couple of minor comments. note that windows CI is currently failing.

jmmease commented:
Thanks @jreback . If I'm not mistaken, It looks like the comments you just made ("blank line", and "issue number") are against an older version of the PR. The new tests were parametrized and moved to `test_sort_values_level_as_str.py`.

Commit:
Added github issue numbers to new tests

jmmease commented:
@jreback The AppVeyor build is passing now, but I'm getting a "The job exceeded the maximum time limit for jobs, and has been terminated." error for most of the Travis builds.

jreback commented:
yeah travis just started failing
not sure why

jreback commented:
rebase and push again, fixed some hanging by Travis CI

Commit:
Merge branch 'master' of https://github.com/pandas-dev/pandas into enh_14353

jmmease commented:
Done. Thanks @jreback . Happy New Year BTW!

jreback commented:
happy new year. will have a final look, but this lgtm. @TomAugspurger

jmmease commented:
Thanks for taking a look @TomAugspurger 
Anything else @jreback ?

jreback commented:
thanks @jmmease really nice patch! keep em coming!

jmmease commented:
Awesome, thanks for all your time on this @jreback @TomAugspurger, @jorisvandenbossche, and @gfyoung !
